backend/src/main/java/pt/ist/socialsoftware/mono2micro/utils/Metrics.java

```diff
@@ -25,38 +25,106 @@ public void calculateMetrics() {
 		float graphCohesion = 0;
 		float graphCoupling = 0;
 
-		for (Cluster cluster : graph.getClusters()) {
-			cluster.setCouplingDependencies(new HashMap<>());
-		}
-		
-		for (Controller controller : graph.getControllers()) {
-			calculateControllerComplexity(controller);
+		List<Controller> graphControllers = graph.getControllers();
+
+		for (Controller controller : graphControllers) {
+			calculateControllerComplexityAndClusterDependencies(controller);
 			graphComplexity += controller.getComplexity();
-			calculateClusterDependencies(controller);
 		}
 
-		graphComplexity /= graph.getControllers().size();
+		graphComplexity /= graphControllers.size();
 		graphComplexity = BigDecimal.valueOf(graphComplexity).setScale(2, RoundingMode.HALF_UP).floatValue();
+
 		this.graph.setComplexity(graphComplexity);
 
-		for (Cluster cluster : graph.getClusters()) {
-			calculateClusterComplexity(cluster);
+		List<Cluster> graphClusters = graph.getClusters();
+
+		for (Cluster cluster : graphClusters) {
+			calculateClusterComplexityAndCohesion(cluster);
 
-			calculateClusterCohesion(cluster);
 			graphCohesion += cluster.getCohesion();
 
 			calculateClusterCoupling(cluster);
 			graphCoupling += cluster.getCoupling();
 		}
 
-		graphCohesion /= graph.getClusters().size();
+		int graphClustersAmount = graphClusters.size();
+
+		graphCohesion /= graphClustersAmount;
 		graphCohesion = BigDecimal.valueOf(graphCohesion).setScale(2, RoundingMode.HALF_UP).floatValue();
 		this.graph.setCohesion(graphCohesion);
-		graphCoupling /= graph.getClusters().size();
+
+		graphCoupling /= graphClustersAmount;
 		graphCoupling = BigDecimal.valueOf(graphCoupling).setScale(2, RoundingMode.HALF_UP).floatValue();
 		this.graph.setCoupling(graphCoupling);
     }
 
+    private void calculateControllerComplexityAndClusterDependencies(Controller controller) {
+		Set<Controller.LocalTransaction> allLocalTransactions = controller.getAllLocalTransactions();
+
+		if (this.controllerClusters.get(controller.getName()).size() == 1) {
+			controller.setComplexity(0);
+
+			// I'm so sorry but... is there a better way?
+			for (Controller.LocalTransaction lt : allLocalTransactions) {
+				// ClusterDependencies
+				Cluster fromCluster = graph.getCluster(lt.getClusterName());
+
+				if (fromCluster != null) { // not root node
+					List<Controller.LocalTransaction> nextLocalTransactions = controller.getNextLocalTransactions(lt);
+
+					for (Controller.LocalTransaction nextLt : nextLocalTransactions) {
+						String toEntity = nextLt.getClusterAccesses().get(0).getEntity();
+						fromCluster.addCouplingDependency(nextLt.getClusterName(), toEntity);
+					}
+				}
+			}
```

Comment:
Ou muito me engano ou, se o controller  tem size == 1 e complexity = 0, as localTransactions dele são compostas por apenas um cluster, ou seja, o segundo for nunca vai correr e diria que o bloco de calcular as dependencias pode sair